### PR TITLE
Introduce Equivalence

### PR DESCRIPTION
**Pull request by mikearnaldi**
First pass, we should now list all the functions that may benefit a variant based on Equivalence

Unfortunately escaping the need for Equal is actually near to impossible as many internals depend on an overridable concept of strict equality

Note: this PR also change Option , Either and These to compare by value because we really should do so, we lose assignability to Json (unless we want to use an intersection instead of an interface). The types should be serializable but not directly deserializable via JSON.parse, that should be considered domain for `/schema`

### CODE REVIEW

**Comment by mikearnaldi** [comment]
@patroza pls review too

**Comment by patroza** [comment]
Still wrapping my head around it all, but I think the JSON.parse argument is mostly valid, except maybe I would look to provide also a lower level hydration utility. That sits between;
Option/Either/These/Chunk etc. not being directly deserialisable, and using Schema. Mainly because Schema is expensive to define and maintain, and if you don’t want to use it (I’m not one of those obviously) I think you should still have an easy time.

Basically it needs to look if something is object, not null and has a _tag. And if it’s a known tag, restore it to its appropriate object type.

**Comment by mikearnaldi** [comment]
I would be generally against specialising json based on tags, and in some sense reinventing schema. I don't see schema as expensive in any way, this behaviour is in line with Date and we never thought it would be a good idea to specialise for dates. If one scares away from schema they aren't going to use options and either over the wire.

**Comment by patroza** [comment]
I know you and me don’t feel that way about Schema.
In any case, maybe you have a point about not leveraging the data types over the wire.
I would still keep this option in the back of our head once we do come across people fitting in the category i described :)

**Comment by mikearnaldi** [comment]
> I know you and me don’t feel that way about Schema. In any case, maybe you have a point about not leveraging the data types over the wire. I would still keep this option in the back of our head once we do come across people fitting in the category i described :) Revivers exist exactly for this reason I would say, it’s not far fetched. But Schema certainly is the better way.

What I say is basically that optimizing for people in that spot is not useful and would de-optimize proper usage, they aren't many and they will have a heart attack anyway when using fp-ts :)
